src/dmd/backend/cgcse.d

```diff
@@ -85,26 +124,128 @@ struct CSE
 
         if (alignsize >= 16 && TARGET_STACKALIGN >= 16)
         {
-            alignment = alignsize;
+            alignment_ = alignsize;
             STACKALIGN = alignsize;
             enforcealign = true;
         }
     }
-}
 
-// Returns: offset of slot i
-uint CSE_offset(int i)
-{
-    return i * CSE.slotSize;
-}
+    /****
+     * Get range of all CSEs filtered by matching `e`,
+     * starting with most recent.
+     * Returns:
+     *  input range
+     */
+    static auto filter(const elem* e)
+    {
+        struct Range
+        {
+            const elem* e;
+            int i;
 
-// Returns: true if CSE was ever loaded
-bool CSE_loaded(int i)
-{
-    return i < csextab.length &&   // array could be shrunk for non-CSEload entries
-           (csextab[i].flags & CSEload);
+          nothrow:
+
+            bool empty()
+            {
+                while (i)
```

Comment:
I think a `foreach_reverse` would be simpler. When reading code and you see the `while` keyword you have no idea of what it's doing. While when reading the `foreach_reverse` keyword you know that in 99% percent of the cases it's going to iterate a data structure in reverse. The difference here is that with `while` you need to read the whole body but with `foreach_reverse` you only need to read the keyword.

Reply:
On the other hand, many people complain about foreach_reverse. The loop is trivial enough I don't think it matters, and can't see how people would be confused about it.

Reply:
> On the other hand, many people complain about foreach_reverse

Yes, and I'm one of them. I have complained that it's a dedicated keyword. It could have been implemented like `foreach.reverse`, in which case only `foreach` would be a keyword. Or not have any language support at all and instead rely on a library solution like `std.range.retro`.